### ISSUE TITLE
FasterLogIterator documentation inconsistent with implementation

### ISSUE THREAD

**Issue by DavidNepozitek** [microsoft/FASTER#477]
Docstring of `FasterLogScanIterator.WaitAsync(CancellationToken token = default)` states that the method returns `false` when reaching the iterator's `endAddress`; however, that is not reflected in the implementation if I'm not mistaken.

https://github.com/microsoft/FASTER/blob/5501dfb24b079b2601665568bfbffcd54065bec1/cs/src/core/Index/FasterLog/FasterLogIterator.cs#L112-L113

At the current state, in a scenario, when a user wants to iterate through a finite part of a log (not using the `GetAsyncEnumerable()`), they must check the `endAddress` manually. That is because (if the following part of the documentation is true) receiving `false` from `GetNext()` doesn't guarantee reaching the `endAddress`.
https://github.com/microsoft/FASTER/blob/5501dfb24b079b2601665568bfbffcd54065bec1/docs/_docs/40-fasterlog-basics.md#L117-L118

In conclusion, the implementation of `WaitAsync` should reflect its docstring, or the docstring should be edited.

**Comment by badrishc** [comment]
I'm not sure what the inconsistency is. The asynchronous call of `await WaitAsync(...)` indeed returns false only when we reach the end of iteration. In a scenario where we are waiting for a page or commit and we have not reached `endAddress`, the `await WaitAsync` call will continue to wait and the `await` will not return.

`GetNext` is a different, and synchronous, API. Being synchronous, it cannot wait on someone else to add items, so it will immediately return false if there isn't data readily available to serve (or you have reached the end of iteration).

One way to use them together is to try getting an item via `GetNext`. If it returns false, and you have not reached endAddress, that means you need to wait for more data. Then, you would call `WaitAsync` to wait for such data to arrive.

**Comment by DavidNepozitek** [comment]
The inconsistency is that the docstring says `WaitAsync` returns false when reaching the `endAddress`; however when you reach the `endAddress` and call `WaitAsync` you won't receive false. 

If I haven't missed anything in the implementation of `WaitAsync`, this method returns false only when the iterator is disposed (as a result of private methods `SlowWaitAsync` or `SlowWaitUncommittedAsync`). There is no check for the `endAddress` in the implementation.

I wrote the second part of my original post just as an example when the semantics of `WaitAsync` is important. I know how to combine these two methods to get a working iteration. However, if I relied on the docstring I wouldn't check whether i reached the `endAddress` and that would result in `WaitAsync` never returning.



I hope that it is clear now, if not I can provide a minimal example of the problem. But it is really just about the following part of the docstring: "false if there will never be more data (log has been shutdown / __iterator has reached endAddress)__"

**Comment by badrishc** [comment]
Got it. Implementation is fixed in attached PR.
